core/shared/src/main/scala/zio/Semaphore.scala

```diff
@@ -118,15 +118,19 @@ object Semaphore {
             }
 
         def restore(promise: Promise[Nothing, Unit], n: Long)(implicit trace: Trace): UIO[Any] =
-          ref.modify {
-            case Left(queue) =>
-              queue
-                .find(_._1 == promise)
-                .fold(releaseN(n) -> Left(queue)) { case (_, permits) =>
-                  releaseN(n - permits) -> Left(queue.filter(_._1 != promise))
-                }
-            case Right(permits) => ZIO.unit -> Right(permits + n)
-          }.flatten
+          promise.isDone.flatMap { done =>
+            if (done) {
+              // If the promise is done, then the permits were acquired, and we need to release them
+              releaseN(n)
+            } else {
```

Comment:
I think this is a dead branch. AFAICT the `release` method is only called after `promise.await`. Did you see this branch being ran?

Reply:
Would it be called if the `promise.await` is interrupted? 

Or would `isDone` return true in that case?

Reply:
looking at the implemenation it looks like `isDone` ould return true. 

So this is incorrect. I need to also check to make sure that the result succeeded, and wasn't interrupted.

Reply:
sorry maybe I was unclear with my code selection. I meant that `isDone` will _always_ return true